clients/python/src/model_registry/_client.py

```diff
@@ -75,16 +81,99 @@ def register_model(
         model_format_name: str,
         model_format_version: str,
         version: str,
+        storage_key: str | None = None,
+        storage_path: str | None = None,
+        bucket_name: str | None = None,
+        bucket_endpoint: str | None = None,
+        bucket_region: str | None = None,
+        service_account_name: str | None = None,
+        author: str | None = None,
+        description: str | None = None,
+        metadata: dict[str, ScalarType] | None = None,
+    ) -> RegisteredModel: ...
+
+    @overload
+    def register_model(
+        self,
+        name: str,
+        *,
+        model_format_name: str,
+        model_format_version: str,
+        version: str,
+        storage_key: str,
+        storage_path: str,
+        bucket_name: str,
+        service_account_name: str | None = None,
+        author: str | None = None,
+        description: str | None = None,
+        metadata: dict[str, ScalarType] | None = None,
+    ) -> RegisteredModel: ...
+
+    @overload
+    def register_model(
+        self,
+        name: str,
+        *,
+        model_format_name: str,
+        model_format_version: str,
+        version: str,
+        storage_key: str,
+        storage_path: str,
+        bucket_name: str,
+        bucket_endpoint: str,
+        bucket_region: str,
+        service_account_name: str | None = None,
         author: str | None = None,
         description: str | None = None,
+        metadata: dict[str, ScalarType] | None = None,
+    ) -> RegisteredModel: ...
+
+    @required_args(
+        # non S3
+        ("uri",),
+        # S3 only
+        (  # pre-configured env
+            "storage_key",
+            "storage_path",
+            "bucket_name",
+        ),
+        (  # custom env or non-default bucket
+            "storage_key",
+            "storage_path",
+            "bucket_name",
+            "bucket_endpoint",
+            "bucket_region",
+        ),
+    )
+    def register_model(
```

Comment:
Rather than having exploded version of each storage type in the overloaded method, could we do 
``` 
s3 = build_s3(); // builds a s3 object from ENV variables etc. or build manaually

register_model(self, name:str ...., s3{bucket_name, bucket_endpoint, ...})
register_model(self, name:str ...., URI)
register_model(self, name:str ...., somethingelse)

```

Reply:
Technically yes, but I personally think the proposed solution makes the API nicer on the user. What would motivate such a change?

<!--
I believe that the interface should fit the user needs instead of us imposing, possibly, a temporary fix to their usage.
In the use-case of notebooks with data connections, I would also argue it's easier to simply provide the default bucket name instead of
```py
from model_registry.utils import build_s3

s3 = build_s3()
```
before registering any models, which just adds to the boilerplate that we're already making such an effort to avoid.
Specially while we're still on beta, I would prefer to keep the interface very small too.
-->

Reply:
Respectfully, what you have IMO is very brittle API. Everytime if a new attribute for any reason need to be added for storage all the API will break(unless it has a default), or a new storage type added the API need to be expanded with more overloaded methods.

I can't say for python style, but If this was in java, I probably would have designed something like

```
interface storage {}
class StorageUri implements storage {}
class S3Storage implements storage {}
class OCIStorage implements storage {}

func storageBuilder():storage {
}

register_model(self, name:str ...., func storageBuilder{})

```
Another advantage is it can give way for functional programming. or no need for if/else logic in the method. above is just example, there can be better variations @tarilabs or @dhirajsb can suggest probably? since this is client facing API I want to make sure we have it correctly defined.

Reply:
>uri was required by design for register_model method
>there must NOT be "any magic" and everything need be explicit

Then we should have `URI` as a parameter in the method signature IMO. very explicit. How one builds the `URI` is irrelevant at this level. my suggestions from above taken as how to build a URI for various storage types

>An alternative to this PR current proposal could be to leave register_model untouched, as that provides 1:1 >mapping with the logical model, and then have specialized (wrapper) methods, ~like register_model_from_uri, >register_model_from_s3, etc, focusing on the actual "storage domains":

agree that this variation provides a more defined API, my view is this is another variation of the proliferation of the API, please note we need to support whatever we propose for a long time.

>less risk of having a property being specialized twice (bucket_path as part of uri and storage_path of kserve, etc)

This is KServe and S3 specific correct? can this be inferred from `URI` and query parameters? If not, force the user to provide them as query parameters by failing the validation check if it is S3?

>naturally one can always decide to consume purely the logical model -oriented register_model as is today

This I can certainly say no :) as we know abstract layers are designed to make developer's life easier.

Reply:
I would personally go with Matteo's suggestion as that would fit easily. However, the S3 URI builder still has at least two distinct overloads, namely for the cases where the user has:

1. default bucket on a pre-configured environment
2. custom bucket/custom environment

> I can't say for python style, but If this was in java, I probably would have designed something like
> 
> ```
> interface storage {}
> class StorageUri implements storage {}
> class S3Storage implements storage {}
> class OCIStorage implements storage {}
> 
> func storageBuilder():storage {
> }
> 
> register_model(self, name:str ...., func storageBuilder{})
> ```

Now, while this could work

> Another advantage is it can give way for functional programming. or no need for if/else logic in the method. above is just example, there can be better variations @tarilabs or @dhirajsb can suggest probably? since this is client facing API I want to make sure we have it correctly defined.

Functional programming is rather uncommon in python due to general and purposeful inefficiencies in the interpreter (https://stackabuse.com/functional-programming-in-python/#somethingstoconsider), so I wouldn't take that into consideration.

Rather, APIs in Python tend to be as declarative as possible, although I'd probably make more sense to not have such rich wrapper methods in that case.

Reply:
I've just updated the PR with the "simplest" version of such a builder for S3. I would say we go for it and leave API abstractions for later, wdyt?

Reply:
I like this variation much better. +1

if you want you can log "info" level messages when endpoint, path or any other optional parameters needed for automation is missing.